python/nano/src/bigdl/nano/pytorch/strategies/ddp_spawn.py

```diff
@@ -0,0 +1,297 @@
+            log.debug(f"[Process {i}]: using OMP_NUM_THREADS: {os.environ['OMP_NUM_THREADS']}")
+            error_queue = mp.SimpleQueue()
+            process = mp.Process(
+                target=_wrap,
+                args=(self._wrapping_function, i, args, error_queue),
+                daemon=False,
+            )
+            process.start()
+            error_queues.append(error_queue)
+            processes.append(process)
+
+        context = ProcessContext(processes, error_queues)
+
+        while not context.join():
+            pass
+
+        os.environ["KMP_AFFINITY"] = init_KMP_AFFINITY
+        os.environ["OMP_NUM_THREADS"] = init_OMP_NUM_THREADS
+
+        # restore the state of child process
+        spawn_output = return_queue.get()
+        self._recover_results_in_main_process(spawn_output, trainer)
+        return spawn_output.trainer_results
+
+
+class DDPSpawnStrategy(_DDPSpawnStrategy):
+    """Extending DDPSpawnStrategy to support launch subprocesses with optimized env variables."""
+
+    strategy_name = "ddp_spawn"
+
+    def __init__(
+        self,
+        num_processes: int = 1,
+        cpu_for_each_process: Optional[List[List[int]]] = None,
+        use_ipex=False,
+        enable_bf16=False,
+        **kwargs: Any
+    ):
+        """Create a DDPSpawnStrategy, adding a cpu_for_each_process parameter."""
+        device = ipex_device() if use_ipex and TORCH_VERSION_LESS_1_10 else 'cpu'
+        parallel_devices = [torch.device(device) for _ in range(num_processes)]
+        cluster_environment = LightningEnvironment()
+
+        if use_ipex and TORCH_VERSION_LESS_1_10 and 'accelerator' not in kwargs:
+            super().__init__(accelerator=create_IPEXAccelerator(),
+                             parallel_devices=parallel_devices,
+                             cluster_environment=cluster_environment, **kwargs)
+        else:
+            super().__init__(parallel_devices=parallel_devices,
+                             cluster_environment=cluster_environment, **kwargs)
+        self.cpu_for_each_process = cpu_for_each_process
+        self.is_distributed = True
+        self.use_ipex = use_ipex
+        self.enable_bf16 = enable_bf16
+
+    def _configure_launcher(self):
+        self._launcher = _DDPSpawnLauncher(self)
+
+    def setup(self, trainer: "pl.Trainer") -> None:
+        """Setup the distributed environment of sub processes, we add ipex optimization here."""
+        invalidInputError(self.model is not None, "You must specify the model.")
+
+        # when using spawn, multiple child processes may update the weights of
+        # the same model, so we should copy the model to avoid it
+        if self.strategy_name == "ddp_spawn":
+            self.model = copy.deepcopy(self.model)
+            # `copy.deepcopy(self.model)` can't copy `self.model.trainer` correctly sometimes,
+            # so we reuse the original trainer
+            self.model.trainer = trainer    # type: ignore
```

Comment:
does trainer also hold a instance of model?

Reply:
In pytorch_lightning 1.6, a trainer holds a strategy, a strategy holds a model, a model hold a trainer.

After assigning to `self.model`, `trainer.model` will refer to new model automatically, because the `trainer` holds a strategy(`self` here), `trainer.model` equals to `self.model`

Reply:
maybe add some explanations in the comments